### PR TITLE
WIP: Syncup with develop branch for config update with multiple model files

### PR DESCRIPTION
**Pull request by hzhao-github**
From Added tests for config update with multiple model files https://github.com/oracle/weblogic-kubernetes-operator/pull/1556

### CODE REVIEW

**Comment by hzhao-github** [comment]
Two Jenkins jobs submitted

File: integration-tests/src/test/java/oracle/kubernetes/operator/ItModelInImageConfigUpdate.java [hunk 1]
```diff
@@ -425,6 +425,179 @@ public void testMiiConfigUpdateJdbcImage() {
     LoggerHelper.getLocal().log(Level.INFO, "SUCCESS - " + testMethodName);
   }
 
+  /**
+   * Create a domain without a JDBC DS using model in image and having configmap
```
Comment (rjeberhard) [inline integration-tests/src/test/java/oracle/kubernetes/operator/ItModelInImageConfigUpdate.java:429]: I'm not yet understanding the description of the initial state.  If the domain configuration does not have a JDBC DS, why is there a need for a config map?  What part of the model, if any, is in the config map at the beginning?

File: integration-tests/src/test/java/oracle/kubernetes/operator/ItModelInImageConfigUpdate.java [hunk 2]
```diff
@@ -425,6 +425,179 @@ public void testMiiConfigUpdateJdbcImage() {
...
+   * in the domain.yaml before deploying the domain. After deploying the domain crd,
```
Comment (rjeberhard) [inline integration-tests/src/test/java/oracle/kubernetes/operator/ItModelInImageConfigUpdate.java:430]: The CRD is created by the operator.  The test will create a domain resource.

File: integration-tests/src/test/java/oracle/kubernetes/operator/ItModelInImageConfigUpdate.java [hunk 3]
```diff
@@ -425,6 +425,179 @@ public void testMiiConfigUpdateJdbcImage() {
...
+
+    try {
+      logTestBegin(testMethodName);
+      LoggerHelper.getLocal().log(Level.INFO,
+          "Creating Domain & waiting for the script to complete execution");
+
+      // create domain w/o JDBC DS using the image created by MII
+      boolean createDS = false;
+      domain = createDomainUsingMii(createDS);
+      Assertions.assertNotNull(domain, "Failed to create a domain");
```
Comment (rjeberhard) [inline integration-tests/src/test/java/oracle/kubernetes/operator/ItModelInImageConfigUpdate.java:454]: import static Assertions.assertNotNull

File: integration-tests/src/test/java/oracle/kubernetes/operator/ItModelInImageConfigUpdate.java [hunk 4]
```diff
@@ -425,6 +425,179 @@ public void testMiiConfigUpdateJdbcImage() {
...
+
+      // copy model files that contains JDBC DS to a dir to re-create cm
+      final String destDir = getResultDir() + "/samples/model-in-image-update-lon";
```
Comment (rjeberhard) [inline integration-tests/src/test/java/oracle/kubernetes/operator/ItModelInImageConfigUpdate.java:457]: What does the "-lon" mean?

File: integration-tests/src/test/java/oracle/kubernetes/operator/ItModelInImageConfigUpdate.java [hunk 5]
```diff
@@ -425,6 +425,179 @@ public void testMiiConfigUpdateJdbcImage() {
...
+              "cm.jdbc.yaml"};
+
+      copyTestModelFiles(destDir, modelFiles);
+
+      // re-create cm to update config and verify cm is created successfully
+      wdtConfigUpdate(destDir, domain);
+
+      // update domain yaml with restartVersion,
+      // apply the domain yaml and verify domain restarted successfully
+      modifyDomainYamlWithRestartVersion(domain, domainNS);
```
Comment (rjeberhard) [inline integration-tests/src/test/java/oracle/kubernetes/operator/ItModelInImageConfigUpdate.java:473]: You're not updating a YAML file, correct?  Won't this method get the current domain, change the restartVersion, and then put it back?  Work with @lennyphan to get a patchDomainResource() method.

Comment (rjeberhard) [reply integration-tests/src/test/java/oracle/kubernetes/operator/ItModelInImageConfigUpdate.java:473]: @lennyphan and @hzhao-github, just realized these tests are in the original test suite.  We'll still need the patchDomainResource() method, but these methods are currently invoking `kubectl patch`

File: integration-tests/src/test/java/oracle/kubernetes/operator/ItModelInImageConfigUpdate.java (hunk 1 above)
Comment (hzhao-github) [reply integration-tests/src/test/java/oracle/kubernetes/operator/ItModelInImageConfigUpdate.java:429]: The model file when creating the domain first looks like this:

domainInfo:
    AdminUserName: '@@SECRET:__weblogic-credentials__:username@@'
    AdminPassword: '@@SECRET:__weblogic-credentials__:password@@'
    ServerStartMode: 'prod'

topology:
    Name: domain1
    AdminServerName: "admin-server"
    Cluster:
        "cluster-1":
            DynamicServers:
                ServerTemplate:  "cluster-1-template"
                ServerNamePrefix: "managed-server"
                DynamicClusterSize: 5
                MaxDynamicClusterSize: 5
                CalculatedListenPorts: false
    Server:
        "admin-server":
            ListenPort: 7001
    ServerTemplate:
        "cluster-1-template":
            Cluster: "cluster-1"
            ListenPort : '@@PROP:MANSVR_PORT@@'
appDeployments:
    Application:
        myear:
            SourcePath: wlsdeploy/applications/sample_app.ear
            ModuleType: ear
            Target: 'cluster-1'

Comment (hzhao-github) [reply integration-tests/src/test/java/oracle/kubernetes/operator/ItModelInImageConfigUpdate.java:429]: very basic things to create a domain

File: integration-tests/src/test/java/oracle/kubernetes/operator/ItModelInImageConfigUpdate.java (hunk 4 above)
Comment (hzhao-github) [reply integration-tests/src/test/java/oracle/kubernetes/operator/ItModelInImageConfigUpdate.java:457]: loading order by name :( let me change it to more clearer one

File: integration-tests/src/test/java/oracle/kubernetes/operator/ItModelInImageConfigUpdate.java (hunk 3 above)
Comment (hzhao-github) [reply integration-tests/src/test/java/oracle/kubernetes/operator/ItModelInImageConfigUpdate.java:454]: k

File: integration-tests/src/test/java/oracle/kubernetes/operator/ItModelInImageConfigUpdate.java (hunk 2 above)
Comment (hzhao-github) [reply integration-tests/src/test/java/oracle/kubernetes/operator/ItModelInImageConfigUpdate.java:430]: k. let me change

Comment (hzhao-github) [reply integration-tests/src/test/java/oracle/kubernetes/operator/ItModelInImageConfigUpdate.java:430]: changed to After the domain resources is created,

File: integration-tests/src/test/java/oracle/kubernetes/operator/ItModelInImageConfigUpdate.java (hunk 3 above)
Comment (hzhao-github) [reply integration-tests/src/test/java/oracle/kubernetes/operator/ItModelInImageConfigUpdate.java:454]: fixed

File: integration-tests/src/test/java/oracle/kubernetes/operator/ItModelInImageConfigUpdate.java (hunk 4 above)
Comment (hzhao-github) [reply integration-tests/src/test/java/oracle/kubernetes/operator/ItModelInImageConfigUpdate.java]: fixed

File: integration-tests/src/test/java/oracle/kubernetes/operator/ItMiiConfigUpdateCm.java [hunk 6]
```diff
@@ -0,0 +1,366 @@
...
+
+    try {
+      logTestBegin(testMethodName);
+      LoggerHelper.getLocal().log(Level.INFO,
+          "Creating Domain & waiting for the script to complete execution");
+
+      // create domain w/o JDBC DS using the image created by MII
+      boolean createDS = false;
+      domain = createDomainUsingMii(createDS, domainNS, testClassName);
+      assertNotNull(domain, "Failed to create a domain");
```
Comment (maggiehe00) [inline integration-tests/src/test/java/oracle/kubernetes/operator/ItMiiConfigUpdateCm.java:120]: I'm not sure for the created domain  assertNotNull would be enough. How about domain.verifyDomainCreated() ?

Comment (hzhao-github) [reply integration-tests/src/test/java/oracle/kubernetes/operator/ItMiiConfigUpdateCm.java:120]: domain.verifyDomainCreated()  is called in MiiBaseTest.java, createMiiDomainWithConfigMap. assertNotNull is just a second layer of check. I know it makes it harder to find if we have layers of extended class.

File: integration-tests/src/test/java/oracle/kubernetes/operator/utils/Secret.java [hunk 7]
```diff
@@ -58,6 +59,55 @@ public Secret(String namespace, String secretName, String username, String passw
     LoggerHelper.getLocal().log(Level.INFO, "command result " + result.stdout().trim());
   }
 
+  /**
+   * Construct secret.
+   * @param namespace namespace
```
Comment (maggiehe00) [inline integration-tests/src/test/java/oracle/kubernetes/operator/utils/Secret.java:64]: How about something like " @param namespace namespace where the secret is going to create"

File: integration-tests/src/test/java/oracle/kubernetes/operator/MiiConfigUpdateBaseTest.java [hunk 8]
```diff
@@ -0,0 +1,659 @@
...
+      TestUtils.createConfigMap(cmName, fileOrDirPath, domainNS, label);
+    } catch (Exception ex) {
+      ex.printStackTrace();
+      fail("Failed to create cm.\n", ex.getCause());
+    }
+  }
+
+  /**
+   * Create a new or update an existing image using model in image.
+   * @param domainMap a Java Map object that storing key and value pairs used to create the Domain
```
Comment (maggiehe00) [inline integration-tests/src/test/java/oracle/kubernetes/operator/MiiConfigUpdateBaseTest.java:245]: "stores" or "object storing"

File: integration-tests/src/test/java/oracle/kubernetes/operator/ItMiiConfigUpdateCm.java [hunk 9]
```diff
@@ -0,0 +1,366 @@
...
+      // verify that JDBC DS is created by checking DS name, JNDI name and read timeout
+      LoggerHelper.getLocal().log(Level.INFO, "Verify that JDBC DS is created");
+      final String[] jdbcResourcesToVerify =
+          {"datasource.name.1=" + dsName,
+              "datasource.jndiname.1=array(java.lang.String,['" + jndiName + "'])",
+              "datasource.readTimeout.1=" + readTimeout_2};
+      for (String prop : jdbcResourcesToVerify) {
+        assertTrue(jdbcResourceData.contains(prop), prop + " is not found");
+        LoggerHelper.getLocal().log(Level.INFO, prop + " exists");
+      }
```
Comment (maggiehe00) [inline integration-tests/src/test/java/oracle/kubernetes/operator/ItMiiConfigUpdateCm.java:140]: It looks that this JDBC DS verification logic can be an util method.

Comment (hzhao-github) [reply integration-tests/src/test/java/oracle/kubernetes/operator/ItMiiConfigUpdateCm.java:140]: The verification was in an util method, But mark likes to it to be clearly assert this way. Let's wait for Mark's comments here before I change it back to use an util method

File: integration-tests/src/test/java/oracle/kubernetes/operator/utils/Secret.java (hunk 7 above)
Comment (hzhao-github) [reply integration-tests/src/test/java/oracle/kubernetes/operator/utils/Secret.java:64]: good idea. Will fix

File: integration-tests/src/test/java/oracle/kubernetes/operator/MiiConfigUpdateBaseTest.java (hunk 8 above)
Comment (hzhao-github) [reply integration-tests/src/test/java/oracle/kubernetes/operator/MiiConfigUpdateBaseTest.java:245]: will fix

Comment (hzhao-github) [reply integration-tests/src/test/java/oracle/kubernetes/operator/MiiConfigUpdateBaseTest.java]: fixed

File: integration-tests/src/test/java/oracle/kubernetes/operator/utils/Secret.java (hunk 7 above)
Comment (hzhao-github) [reply integration-tests/src/test/java/oracle/kubernetes/operator/utils/Secret.java]: fixed

**Comment by hzhao-github** [comment]
after syncup with develop b ranch, all four tests passed in my local

oracle.kubernetes.operator.ItMiiConfigUpdateCm
[INFO] 
[INFO] Results:
[INFO] 
[INFO] Tests run: 3, Failures: 0, Errors: 0, Skipped: 0

oracle.kubernetes.operator.ItMiiConfigUpdateImage
[INFO] 
[INFO] Results:
[INFO] 
[INFO] Tests run: 2, Failures: 0, Errors: 0, Skipped: 0

oracle.kubernetes.operator.ItMiiConfigUpdateMulti
[INFO] 
[INFO] Results:
[INFO] 
[INFO] Tests run: 2, Failures: 0, Errors: 0, Skipped: 0

oracle.kubernetes.operator.ItMiiConfigUpdateSecret
[INFO] 
[INFO] Results:
[INFO] 
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0

**Pull request closed without merging**